ES2018 RegExp enhancements

mysticatea opened the pull request:
- [X] Replaces `RegExp` validation by own validator. The validator satisfies the full-spec regular expressions based on [A.8 Regular Expressions](https://www.ecma-international.org/ecma-262/8.0/#sec-regular-expressions) and [B.1.4 Regular Expressions Patterns](https://www.ecma-international.org/ecma-262/8.0/#sec-regular-expressions-patterns). The validator generates the same error messages as Node.js native `RegExp` implementation. (I checked the code coverage of the validator by [nyc](https://github.com/istanbuljs/nyc) and made almost 100%)
- [X] Enhances the validator to support RegExp named capture groups (https://github.com/tc39/ecma262/commit/95ec0c6e67af8e6294225e090613bf4e87fed446)
- [x] Enhances the validator to support RegExp Unicode Property Escapes (https://github.com/tc39/ecma262/commit/0ae3582a32125f560ecac540c5444a05e92e5a41)
- [x] Enhances the validator to support RegExp Lookbehind Assertions (https://github.com/tc39/ecma262/commit/bf8a9be462d76cf86b5906e3ccad6a91568199f9)

I'd like to get advice for the direction.

marijnh commented:
This is great. One thing I'm a bit wary of is the code weight. This is written in a very, err, software-engineery way. The RegExp grammar isn't all that complicated, and I figure if we cut down some of the abstraction (numeric literals rather than constants, etc) this could be at least two times smaller. If you want i can take a stab at that.

Also, it may, though it makes little sense from a functionality perspective, be a good idea to make these parser methods instead of plain functions, so that plugins can add RegeExp features.

mysticatea commented:
Thank you.

Yes, it's not slim. There are two reasons.

- One is that my brain couldn't keep ASCII code in my memory and I want input completion to help me. I was not bothered about the constants since minification tools can fold those.
- Another one is that I intended to make the validator's structure closing to the spec's structure as possible. I thought that it will help us to make enhancement easily in future since TC39 proposals are differences to the current spec.

> Also, it may, though it makes little sense from a functionality perspective, be a good idea to make these parser methods instead of plain functions, so that plugins can add RegExp features.

I will move the functions into the validator.

adrianheine commented:
https://github.com/jviereck/regjsparser and it's tests might be a useful resource for this.

marijnh commented:
> I will move the functions into the validator.

Plugins can not extend this class, though (I mean, I guess they sort of can, but not in a composable way). I was proposing for the methods that might be overridden to live on the `Parser` class.

mysticatea commented:
I see. But I'm afraid it since the validator has a ton of state and methods to validate RegExp patterns. I'm not sure if the merging is better... For enhancement, is `this.regexpValidator` property not enough?

marijnh commented:
> For enhancement, is this.regexpValidator property not enough?

Not really. You should still keep the regexp-related state in a separate object, but put the main parsing functions on `Parser`, passing them this state as an argument.

mysticatea commented:
OK, I will try it.

mysticatea commented:
I moved the ton of methods into `Parser.prototype`. To clarify those methods are to validate RegExp patterns, I added `validateRegExp_` prefix to those methods.

marijnh commented on src/regexp.js:
```diff
@@ -0,0 +1,1124 @@
+  }
+
+  // Node.js 0.12/0.10 don't support String.prototype.codePointAt().
+  codePointAt(i) {
```
Is there any case where it actually matters whether we have a single or two (non-syntax) characters? Could we make do by just looking at code units not code points?

(If not, there may be a problem where this gets regexps without `u` flag wrong, since those _should_ treat surrogate pairs as two characters. If yes, we could maybe drop the extra complexity around code points?)

mysticatea replied:
Thank you for good catch. It was a problem. I fixed it and added tests.

jonnymanok commented:
I noticed this code isn't bullet proof and missing some validation. You can parse `/([a ]\b)*\b/` and it works, but outputs invalid syntax. Invalid WS in this case. There are other cases as well like you need to check for opening brackets, braces / curly and parens. This will parse without an issue even if they are wrong.

mysticatea commented:
@jonnymanok Thank you for the comment. I added a test for `/([a ]\b)*\b/`. It's valid syntax as same as V8.

jonnymanok commented:
@mysticatea  Ok.  BTW. This is valid with and without a unicode flag: `/a-b\uD83D\uDCA9-\uD83D\uDCAB/u` If you wrap it inside brackets like this `/[a-b\uD83D\uDCA9-\uD83D\uDCAB]/u` it will throw if no unicode flag. Isn't this the same thing ?

mysticatea commented:
> BTW. This is valid with and without a unicode flag: /a-b\uD83D\uDCA9-\uD83D\uDCAB/u If you wrap it inside brackets like this /[a-b\uD83D\uDCA9-\uD83D\uDCAB]/u it will throw if no unicode flag. Isn't this the same thing ?

Those are not same. The former is just a string `"a-b💩-💫"`. The latter is one character which is in the range `a` to `b` and `💩` to `💫` if with u flag. But the latter has invalid range `\uDCA9-\uD83D` if without u flag, it's syntax error.
Similar tests are here: https://github.com/acornjs/acorn/pull/673/files#diff-cd0921c3adebdccb0897f36a2d491edeR955

> this should parse /(?<\\u{41}bc\\u{41}>)\\k<\\u{41}bc\\u{41}>/

Yes, it's valid name if the `\\` means `\`. The `\u{41}` is `A`, so the group name is `AbcA`.
Similar test is here: https://github.com/acornjs/acorn/pull/673/files#diff-a972742ff2aaf03cf6fc24c38e847b2eR61

> With that I would have expected this /\p{Script=Greek}/ to throw an syntax error. It's missing the unicode flag.

No. It's valid. It just matches to the string `"p{Script=Greek}"`. You can try it in devtools' console.

mysticatea commented:
The validator in this PR is based on V8 behavior, but the behavior might not be formal spec about lookbehind assertions: https://github.com/tc39/ecma262/pull/1102

marijnh commented:
This is great work. It does add more than 25% to the byte size of `acorn.js`. @RReverser @adrianheine Is that an acceptable cost for this feature?

mysticatea commented:
Sorry, I made some changes:

- 2b229f0 fixed Lookbehind assertions to follow tc39/ecma262#1102.
- 307e347, df4b452, and ea163cb make shorter a bit.

marijnh commented:
I've gone ahead and merged this as 0d20f67. Great work.

mysticatea commented:
Thank you!

adrianheine commented:
Thanks for implementing this, @mysticatea, and thanks for reviewing, @marijnh! Running the regjsparser test suite on acorn, I found four failures:

1. `/[\B]/` should not parse, but does
2. Three test cases test something like `((([a]+)?/?>?)?)`. I don't understand why that should pass. @mathiasbynens, could you give me a hint?

mathiasbynens commented:
cc @jviereck

mathiasbynens commented:
That looks like a mistake to me. cc @jviereck

mysticatea commented:
@adrianheine `/[\B]/` should be parsed because `\B` is equivalent to `B` in Annex B. (https://www.ecma-international.org/ecma-262/8.0/#prod-annexB-IdentityEscape)

jviereck commented:
> Three test cases test something like ((([a]+)?/?>?)?). I don't understand why that should pass. @mathiasbynens, could you give me a hint?
> Re: 2, that looks like a mistake to me. cc @jviereck

I tried to figure out what you are referring to (is there a test in regjsparse that tests for this sequence) but couldn't find/figure it out. Happy to take a look once I know what you refer to. Thanks :)

adrianheine commented:
@jviereck line 1254 in test-data.json for example

@mysticatea Right, regjsparser doesn't follow Annex B, I actually discussed that in jviereck/regjsparser#90 but forgot about it again.

mysticatea commented:
About `((([a]+)?/?>?)?)`, `A.8 Regular Expressions` syntax allows it, so `new RegExp("((([a]+)?/?>?)?)")` is valid. However, [RegularExpressionLiteral](https://www.ecma-international.org/ecma-262/8.0/#prod-RegularExpressionLiteral) production stops at the `/` in the pattern, so the RegExp literal `/((([a]+)?/?>?)?)/` is syntax error.

jviereck commented:
> Three test cases test something like ((([a]+)?/?>?)?). I don't understand why that should pass. @mathiasbynens, could you give me a hint?

When I implemented regexp.js (where regjsparse.js emerged from) and I imported the test262 tests, I recall there were some RegExps matching XML that were failing. It looked like these tests in test262 should fail from reading the spec, but they did not and also browsers were happy. So, I recall adjusting the parser to make things work.

mysticatea commented:
@marijnh Do you have a plan to publish a new version?

marijnh commented:
Do you have time to look into the performance regression reported [here](https://github.com/acornjs/acorn/issues/665#issuecomment-368390784) first? That seems a lot for just parsing regexps, which should be only a small fraction of the source code by size—maybe if you profile something obvious will stand out.

jonnymanok commented:
@marijnh I did some research regarding the performance loss and found out that the performance loss is only existing for libraries that rely on heavy use of regular expressions such as jQuery and Angular. So the perf loss isn't existing for libs that uses a small amount of regexp. And also parsing out a single regexp or two does not generate any perf loss. Hope that help :)

marijnh commented:
> the performance loss is only existing for libraries that rely on heavy use of regular expressions such as jQuery and Angular

Looking through the source code for jquery I don't see all that many regular expressions, so I don't understand why validating them would produce such a noticeable slowdown.

> I may think the issue or the bottle neck is where you create a new regexp prototype instance when validation regular expressions

This was done before the recent patches as well, and is required by the ESTree spec, so that's not the source of the regression and not something we can change.

jonnymanok commented:
@marijnh I looked into the regexp validation code. Not sure how to reproduce it, but you can use `console.log`. It looks like the code for indexing the chars are run multiple times even if they shouldn't, and there is used a "length check" to return if nothing to be done.  Several places I saw the same code even for a single regexp.

I also noticed in top of the script it reparse for some regular expressions. As far as I understand this code was modeled after V8, so I did a comparison and there is no reparsing there. Not sure if this is enough to do a perf regression. 

It also seems to - (*not 100% sure*) that there are a few "code validation duplication" when parsing out unicode e.g. surrogate pairs.

mysticatea commented:
In my env (Node 8.9.3, Windows 10), performance impact looks small.

```js
"use strict"

const fs = require("fs")
const {performance} = require("perf_hooks")
const {parse} = require("./dist/acorn")
const code = fs.readFileSync("jquery-3.3.1.js", "utf8")

const times = []
for (let i = 0; i < 0x2FF; ++i) {
  const t0 = performance.now()
  parse(code, {ecmaVersion: 8})
  const t1 = performance.now()

  process.stdout.write(".")
  times.push(t1 - t0)
}
times.sort()
console.log("\nMEAN:", times[times.length / 2 | 0], "ms")
```

Before: `MEAN: 20.827545000240207 ms` (ba939ea353d92d90cdbed2dc6f4b28082d364e40)
After: `MEAN: 21.117823000997305 ms` (0d20f67b03e0784bdd18eb88e0f76e6a53043404) (+1.4%)

jonnymanok commented:
@mysticatea you are only testing jquery. Run acorn own benchmark and you will notice it. Also try benchmarking librares such as angular 1.6, jquery mobile, ts etc. Larger the libs are and more regular expressions, the perf drop compared to current scorn version.

mysticatea commented:
I tried with jquery and jquery-mobile:

Before: `MEAN: 80.36216400004923 ms` ms (ba939ea)
After: `MEAN: 80.5539809986949 ms` (0d20f67) (+0.2%)

```js
"use strict"

const fs = require("fs")
const {performance} = require("perf_hooks")
const {parse} = require("../dist/acorn")
const code = [
  fs.readFileSync("jquery-3.3.1.js", "utf8"),
  fs.readFileSync("jquery.mobile-git.js", "utf8")
].join("\n")

const times = []
for (let i = 0; i < 0x3FF; ++i) {
  const t0 = performance.now()
  parse(code, {ecmaVersion: 8})
  const t1 = performance.now()

  process.stdout.write(".")
  times.push(t1 - t0)
}
times.sort()
console.log("\nMEAN:", times[times.length / 2 | 0], "ms")
```

jonnymanok commented:
@mysticatea [here](https://marijnhaverbeke.nl/acorn/test/bench/index.html)

mysticatea commented:
@jonnymanok Thanks. But how I can run it in my local with specific commit?

jonnymanok commented:
@mysticatea Compile it, and modify the source to run your locale copy. After DL the source from this repo.

Here is my current result on my computer. Please note that the results are different from computer to computer. 
![new_bench](https://user-images.githubusercontent.com/35224185/36709836-07ea58e6-1bb5-11e8-9f64-5d53a2adb688.png)

mysticatea commented:
Ah, I find it in `test/bench` directory. Thanks.

jonnymanok commented:
NP. And here is a screenshot from another computer I ran this on. I compared Acorn dev against both Esprima and the cherow parser. I still notice a perf regression both in jquery and react library.
![bench1](https://user-images.githubusercontent.com/35224185/36710097-d23594fc-1bb6-11e8-957f-586d3417de87.png)

mysticatea commented:
I tried it with all commits between the current master and 5.4.1.

![image](https://user-images.githubusercontent.com/1937871/36710623-81357156-1bc2-11e8-85cc-6c19bfcbbecc.png)

mysticatea commented:
The bench looks to show different results on each try.

![image](https://user-images.githubusercontent.com/1937871/36710731-2c010abe-1bc3-11e8-8686-2bdfc8bce6da.png)

mysticatea commented:
![image](https://user-images.githubusercontent.com/1937871/36710883-eec5b2b6-1bc3-11e8-8c20-9120fa9339f7.png)

mysticatea commented:
I didn't find performance regression... 🤔

mysticatea commented:
In my env, the dev seems faster than 5.4.1.
Mysterious.

![image](https://user-images.githubusercontent.com/1937871/36711084-cf7a8cb4-1bc4-11e8-8ddc-4b2cfb77b78e.png)

jonnymanok commented:
I ran out of time in my end, but I notice you didn't find much regression. Try to add Angular 1.6, typescript and jquery mobile. And also note that the results differ from computer to computer. 
I ran this on a i5 7th gen laptop with windows 10, and a i3 4th gen desktop w/ Windows Vista

KFlash commented:
Nothing obvious  in current code that cause a huge perf loss. It's room for optimization and performance tweaks, but that's it. I tested this against jslint. [Line 423](https://github.com/douglascrockford/JSLint/blob/master/jslint.js#L423) contain a very long regexp you can validate against.

marijnh commented:
Okay, that sounds like there's no real problem. I've released 5.5.0

jdalton commented:
Is there a way to disable this and defer to the given engine?
